File: cutstash.py
"""
Takes FASTQ records from STDIN; cuts trailing lowercase sequence and the
associated qualities, stashing them in the sequence id comment field as a
SAM-style tag: QT:Z:[seq_cut]+[qal_cut]. Existing comments are dropped.
Example usage:
	cutadapt --action=lowercase ...|python cutstash.py|bwa mem -C ...
"""
import sys


def cutstash(fh):
    try:
        for line in fh:
            # drop any existing comment
            sid = line.rstrip().split()[0]
            seq = next(fh).rstrip()
            plus = next(fh).rstrip()
            qual = next(fh).rstrip()
            assert sid.startswith('@')
            assert plus.startswith('+')
            # search backwards for efficiency
            pos = end = len(seq) - 1
            while pos >= 0 and seq[pos].islower():
                pos -= 1
            if pos < end:
                seq, seq_trimmed = seq[:pos+1], seq[pos+1:].upper()
                qual, qual_trimmed = qual[:pos+1], qual[pos+1:]
                tag = f'QT:Z:{seq_trimmed}+{qual_trimmed}'
                sid = f'{sid} {tag}'
            # faster than print()
            sys.stdout.write(f'{sid}\n')
            sys.stdout.write(f'{seq}\n')
            sys.stdout.write(f'{plus}\n')
            sys.stdout.write(f'{qual}\n')
    except StopIteration as si:
        sys.stderr.write(f'Unexpected end of FASTQ record at {sid}\n')
        return 2
    return 0


if __name__ == '__main__':
    if len(sys.argv) > 1:
        print(__doc__)
    else:
        sys.exit(cutstash(sys.stdin))

File: cutstash.cxx
#include <iostream>

using namespace std;

int main(int argc, char* argv[]) {

  if (argc > 1) {
    cout << "Takes FASTQ records from STDIN; cuts trailing lowercase sequence and the\n";
    cout << "associated qualities, stashing them in the sequence id comment field as a\n";
    cout << "SAM-style tag: ZT:Z:[seq_cut]+[qal_cut]. Existing comments are dropped.\n";
    cout << "Example usage:\n";
    cout << "\tcutadapt --action=lowercase ...|cutstash|bwa mem -C ...\n";
    return 0;
  }

  // Un-sync and un-tie the I/O; this makes a HUGE difference in speed when
  // reading from stdin and writing to stdout with streams.
  ios::sync_with_stdio(false);
  cin.tie(0);

  string sid, seq, seq_cut, sep, qal, qal_cut;
  int64_t end, pos, space;
  while(getline(cin, sid)) {
     // drop any existing comment
     space = sid.find_first_of(" \t");
     if (space != string::npos){
       sid = sid.substr(0, space);
     }
     getline(cin, seq);
     getline(cin, sep);
     getline(cin, qal);
     end = pos = seq.length() - 1;
     while (pos >= 0 && islower(seq[pos])) { pos--; }
     if (pos < end) {
       seq_cut = seq.substr(pos+1, string::npos);
       for (auto & c: seq_cut) { c = toupper(c); }
       seq = seq.substr(0, pos+1);
       qal_cut = qal.substr(pos+1, string::npos);
       qal = qal.substr(0, pos+1);
       sid = sid + " ZT:Z:" + seq_cut + "+" + qal_cut;
     } 
     cout << sid << "\n";
     cout << seq << "\n";
     cout << sep << "\n";
     cout << qal << "\n";
  }
  return 0;
}
